### PR TITLE
Integrate K8s probes to hit Pebble /v1/health endpoint

### PR DESCRIPTION
**Pull request by benhoyt**
This adds Kubernetes probes in the sidecar pod-spec definition to hit the new Pebble `/v1/health` HTTP endpoint (which aggregates the health of checks). The liveness probe hits `/v1/health?level=alive` and the readiness probe hits `/v1/health?level=ready`. Both endpoints return 200 `{"healthy": true}` if no Pebble checks have been defined -- the charmer must explicitly define checks with `level: alive` and/or `level: ready` to use the feature.

If the level=alive check is not healthy (and there's no level=ready check defined), that implies level=ready is unhealthy too -- the `/v1/health` aggregation takes this semantic meaning of the levels into account.

The intent of this change is to allow charmers to make robust, production-ready charms, with K8s restarting pods if they go unhealthy. See the [spec](https://docs.google.com/document/d/1d6-h3UAt2VPUSvlkVF30l8iuDW8raRNkHbp5M6NUo1A/edit#heading=h.5rw2t2gzo2xf).

The Pebble code changes for health checks and the `/v1/health` API are at:

* https://github.com/canonical/pebble/pull/85
* https://github.com/canonical/pebble/pull/86

## QA steps

These steps create a new microk8s controller, deploy the `snappass-test` charm, and play with the health checks integration:

```
# Deploy a sidecar charm app
make microk8s-operator-update  # to update Docker operator image
juju bootstrap microk8s
juju add-model snappass
juju deploy snappass-test
juju status  # visit http://<address>:5000 to view snappass web app
microk8s kubectl describe pods -n snappass snappass-test-0  # view K8s pod details

# Add a Pebble layer to ignore "redis-server" exiting but also an "alive" check
# write contents of layer.yaml
echo "
services:
    redis:
        override: merge
        on-success: ignore
        on-failure: ignore

checks:
    redis-up:
        override: replace
        level: alive
        tcp:
            port: 6379
" >layer.yaml
# copy layer to charm container
microk8s kubectl cp -n snappass -c charm ./layer.yaml snappass-test-0:/layer.yaml
# run "pebble add" to add layer
microk8s kubectl exec -n snappass snappass-test-0 -c charm -- /bin/sh -c 'PEBBLE_SOCKET=/charm/containers/redis/pebble.socket /charm/bin/pebble add l1 /layer.yaml'
# run "pebble replan" to pick up service changes
microk8s kubectl exec -n snappass snappass-test-0 -c charm -- /bin/sh -c 'PEBBLE_SOCKET=/charm/containers/redis/pebble.socket /charm/bin/pebble replan'

# Kill redis-server so check will go unhealthy after 3 attempts
sudo killall redis-server  # be careful :-)
# monitor Pebble check state
microk8s kubectl exec -n snappass snappass-test-0 -c charm -- /bin/sh -c 'PEBBLE_SOCKET=/charm/containers/redis/pebble.socket /charm/bin/pebble checks'

# After ~30 seconds it should go unhealthy and the K8s probe will fail, terminating the container
# monitor K8s status with:
microk8s kubectl describe pods -n snappass snappass-test-0

# Other useful commands
# run "pebble plan" to show updated plan
microk8s kubectl exec -n snappass snappass-test-0 -c charm -- /bin/sh -c 'PEBBLE_SOCKET=/charm/containers/redis/pebble.socket /charm/bin/pebble plan'
# hit /v1/health endpoint manually
microk8s kubectl exec -n snappass snappass-test-0 -c charm -- curl 'http://localhost:38813/v1/health?level=alive'
# run "pebble services" to show service status
microk8s kubectl exec -n snappass snappass-test-0 -c charm -- /bin/sh -c 'PEBBLE_SOCKET=/charm/containers/redis/pebble.socket /charm/bin/pebble services'
# show Pebble logs via kubectl (can use "-f" to follow)
microk8s kubectl logs -n snappass snappass-test-0 -c redis
```

## Documentation changes

See [here](https://juju.is/docs/sdk/pebble#heading--service-auto-restart). Diff at https://github.com/benhoyt/juju-docs/pull/1.

### CODE REVIEW

File: caas/kubernetes/provider/application/application.go
Comment (hpidcock) [inline caas/kubernetes/provider/application/application.go:62]: I like this :)

File: caas/kubernetes/provider/application/application.go
Comment (hpidcock) [inline caas/kubernetes/provider/application/application.go:1374]: This reminds me we need to finish the caasprober worker to actually pull from the local state of the uniter. Currently liveness, readiness and startup probes just check if the unit agent is alive.

**Review by hpidcock — COMMENTED** [review]
Looking good so far.

**Comment by SimonRichardson** [comment]
Closing this because of no activity, please reopen if you want to move this forward.

**Comment by benhoyt** [comment]
Yeah, we'll still want to merge this (once Pebble health checks are merged), but I can reopen then - realistically it'll be another couple of weeks.

**Comment by benhoyt** [comment]
Reopening now that these changes are merged into Pebble. Will update shortly.

**Review by hpidcock — APPROVED** [review]
Awesome thanks for pushing this through

**Comment by benhoyt** [comment]
$$merge$$

**Comment by benhoyt** [comment]
$$merge$$

**Comment by benhoyt** [comment]
$$merge$$

**Pull request merged**
